Merge feature/refactor-classes

mspaeth opened the pull request:
Refactored cell class and gamefield class,
Implement getOpt in runGame.php to run game with arguments from console.

**Tests done**
- Run game of life (tried runGame.php and index.php) and set a glider

mspaeth committed:
Refactored cell class

mspaeth committed:
Gamefield php has now the cell class included

mspaeth committed:
Implement getOpt

Create runGame.php to run game with getOpt

mbohlaender commented on composer.json:
```diff
@@ -0,0 +1,15 @@
+{
```
please move composer.json and vendor directory to lib/external/

mbohlaender commented on v1/index.php:
```diff
@@ -6,10 +6,10 @@
 require_once "lib/gamefield.php";
 require_once "lib/output.php";
 require_once "lib/GifCreator.php";
+require __DIR__ . '../vendor/autoload.php';
```
Fix this due to changed path of vendor directory

mbohlaender commented on v1/lib/cell.php:
```diff
@@ -45,4 +45,19 @@ public function getCoordX()
+        $this->isAlive = false;
+    }
+
+    public function isAlive()
```
No documentation at all, of course this is not that important. But this is a learning project (!!!), so you need to learn how it works!

mbohlaender commented on v1/lib/cell.php:
```diff
@@ -45,4 +45,19 @@ public function getCoordX()
+        $this->isAlive = true;
+    }
+
+    public function dead()
```
No documentation at all, of course this is not that important. But this is a learning project (!!!), so you need to learn how it works!

mbohlaender commented on v1/lib/cell.php:
```diff
@@ -45,4 +45,19 @@ public function getCoordX()
         return $this->x;
     }
+
+    public function life()
```
No documentation at all, of course this is not that important. But this is a learning project (!!!), so you need to learn how it works!

mbohlaender commented on v1/lib/cell.php:
```diff
@@ -13,13 +13,13 @@ class Cell
+    private $isAlive;
 
     /**
      * Cell constructor.
```
What is this sentence saying?

mbohlaender commented on v1/lib/baseOutput.php:
```diff
@@ -0,0 +1,12 @@
+ * User: Max
+ * Date: 05.02.2016
+ * Time: 09:50
+ */
```
Take a look at how we are documentating a file (take a look at the file I send to you via flowdock). You can set this as preset in your phpstorm. Ask me on how to do it.

mbohlaender commented on v1/lib/baseOutput.php:
```diff
@@ -0,0 +1,12 @@
+ * Date: 05.02.2016
+ * Time: 09:50
+ */
+abstract class BaseOutput
```
There is no class documentation. Describe what this baseclass is doing.

mbohlaender commented on v1/lib/baseOutput.php:
```diff
@@ -0,0 +1,12 @@
+ */
+abstract class BaseOutput
+{
+    abstract public function output(GameFieldController $_gameFieldController, $_cycle);
```
Sorry, but this is just to less. You need to start thinking on how it is for another developer, if he needs to implement a new Output plugin. Describe here, what he needs to implement there.

See thhis screenshot on how this can be done ;-)
![abstractfunctiondocumentation](https://cloud.githubusercontent.com/assets/965802/12843742/9c7ebd32-cbfb-11e5-9fa9-632ea17ba3f7.PNG)

mbohlaender commented on v1/lib/consoleOutput.php:
```diff
@@ -0,0 +1,38 @@
+ * User: Max
+ * Date: 05.02.2016
+ * Time: 09:57
+ */
```
Use correct file documentation here please..

mbohlaender commented on v1/lib/consoleOutput.php:
```diff
@@ -0,0 +1,38 @@
+ */
+require_once 'baseOutput.php';
+
+class ConsoleOutput extends BaseOutput
```
Where is the class documentation?

mbohlaender commented on v1/lib/consoleOutput.php:
```diff
@@ -0,0 +1,38 @@
+ * Date: 05.02.2016
+ * Time: 09:57
+ */
+require_once 'baseOutput.php';
```
rename file baseoutput.php...no camel case in file names please.

mbohlaender commented on v1/lib/consoleOutput.php:
```diff
@@ -0,0 +1,38 @@
+
+class ConsoleOutput extends BaseOutput
+{
+    public function output(GameFieldController $_gameFieldController, $_cycle)
```
Where is the function documentation?

mbohlaender commented on v1/lib/gamefield.php:
```diff
@@ -6,21 +6,44 @@
  * Date: 15.01.2016
  * Time: 12:02
  */
```
Please use correct file documentation.

mbohlaender commented on v1/lib/gamefield.php:
```diff
@@ -6,21 +6,44 @@
  */
+require_once "cell.php";
+
 class GameField
```
Where is the class documentation?

mbohlaender commented on v1/lib/gamefield.php:
```diff
@@ -6,21 +6,44 @@
+
 class GameField
 {
     /** @var  Cell[] */
```
Good :+1:

mbohlaender commented on v1/lib/gamefield.php:
```diff
@@ -6,21 +6,44 @@
 
     /**
      * GameField constructor.
      * @param $_cells cell[] Array which contains instances of cell.
```
:facepunch:

mbohlaender commented on v1/lib/gamefield.php:
```diff
@@ -6,21 +6,44 @@
     }
 
     /**
+     * Private function which creates the array of cells which are the gamefield.
```
Everyone can see, that it is private...

mbohlaender commented on v1/lib/gamefield.php:
```diff
@@ -6,21 +6,44 @@
+            for($j=0; $j<$this->width; $j++)
+            {
+                $cell = new Cell($j,$i,0);
+                $this->cells[] = $cell;
```
Make this two lines in one line:
$this->cells[] = new Cell($j,$i,0);

mbohlaender commented on v1/lib/gamefield.php:
```diff
@@ -6,21 +6,44 @@
+    {
+        // Make columns
+        for($i=0; $i<$this->height; $i++)
+        {
```
Didn't you read the style guide? A for is followed by a space:
for ()

And:
change documenation:
for ()
{ // create columns

mbohlaender commented on v1/lib/gamefield.php:
```diff
@@ -6,21 +6,44 @@
+        {
+            // Make rows
+            for($j=0; $j<$this->width; $j++)
+            {
```
Didn't you read the style guide? A for is followed by a space:
for ()

And:
change documenation:
for ()
{ // create rows

mbohlaender commented on v1/lib/gamefield.php:
```diff
@@ -6,21 +6,44 @@
+        }
+    }
+    /**
      * Necessary for deep cloning the gamefield.
```
What for this is needed?

mbohlaender commented on v1/lib/gamefield.php:
```diff
@@ -78,11 +101,21 @@ public function getNeighboursByCell(Cell $cell)
         return $neighbours;
     }
+
+    public function getHeight()
```
Where is the function documentation?

mbohlaender commented on v1/lib/gamefield.php:
```diff
@@ -78,11 +101,21 @@ public function getNeighboursByCell(Cell $cell)
+        return $this->height;
+    }
+
+    public function getWidth()
```
Where is the function documentation?

mbohlaender commented on v1/lib/output.php:
```diff
@@ -47,7 +47,11 @@ public function runGame($_cycle, $_mode)
                 // Make rows
                 for($j=0; $j<$this->x; $j++)
                 {
```
Your output class is not needed anymore. Please remove it. (save it locally so that you can reuse the code in your new plugins)

mbohlaender commented on v1/runGame.php:
```diff
@@ -0,0 +1,85 @@
+ * User: Max
+ * Date: 05.02.2016
+ * Time: 10:26
+ */
```
Use correct file documentation please

mbohlaender commented on v1/runGame.php:
```diff
@@ -0,0 +1,85 @@
+            $output = new ConsoleOutput();
+            $output->output($gameFieldController, $numCycles);
+        }
+        else die("Missing --numCycles argument");
```
We always this die? Just use echo "BLABLUB\N";
Otherwise you may will have many rewritings if the code should NOT stop here..

mbohlaender commented on v1/runGame.php:
```diff
@@ -0,0 +1,85 @@
+
+
+if ($options->getOption('w')) $x = $options->getOption('h');
+else die("Argument --width is missing");
```
I do not think that this is needed because you are saying above, that this options are required...Please check what getopt is doing, if you do not add this options!

mbohlaender commented on v1/runGame.php:
```diff
@@ -0,0 +1,85 @@
+            $numCycles = $options->getOption('c');
+
+            $gameFieldController = new GameFieldController($gameField);
+            $output = new ConsoleOutput();
```
and now we are getting damn generic:
Please write something like $outputClass = $options-getOption("c")."Output"; ($options-getOption("c") must return Console", than build build output like  $output = new $outputClass();

Than you need to move all your output plugins (just ConsoleOutput so far) into lib/output/  and then you need to include all files of this directory:
http://stackoverflow.com/questions/14595072/include-all-directory-files-in-class

mbohlaender commented on v1/runGame.php:
```diff
@@ -0,0 +1,85 @@
+            new Getopt\Option('w','width',Getopt\Getoptv2::REQUIRED_ARGUMENT),
+            new Getopt\Option('h','height',Getopt\Getoptv2::REQUIRED_ARGUMENT),
+            new Getopt\Option('c','numCycles',Getopt\Getoptv2::REQUIRED_ARGUMENT),
+            new Getopt\Option('t','type',Getopt\Getoptv2::REQUIRED_ARGUMENT)
```
Add a help option and output some help. Take a look at the documentation AND on this what I send you via flowdock.

mbohlaender commented:
Somehow i'm not able to comment the GamefieldController. Please also fix all documentation stuff in this class...

mspaeth committed:
Moved composer.json to /lib/external  #GoL work 2m

mspaeth committed:
Comment cell.php #GoL work 10m

mspaeth committed:
Comment baseOutput.php and renamed $cycles to $numCycles #GoL work 15m

mspaeth committed:
Renamed baseOutput.php to baseoutput.php #GoL work 1m

mspaeth committed:
Comment consoleOutput.php

mspaeth committed:
Comment gamefield.php #GoL work 15m

mspaeth committed:
Remove index.php and output.php #GoL work 1m

mspaeth committed:
Fixed file documentation #GoL work 5m

mspaeth committed:
Change die to echo in runGame.php and add file documentation to gamefieldcontroller.php #GoL work 15

mspaeth committed:
Change getop version from 3.x to 1.4.1, add help text #GoL work 40m

mspaeth committed:
Remove spaces #GoL work 1m

mbohlaender commented on .gitignore:
```diff
@@ -0,0 +1,2 @@
+vendor
```
Why didn't you adapt it to the new place?

mbohlaender commented on v1/lib/baseOutput.php:
```diff
@@ -0,0 +1,23 @@
+<?php
```
why you didn't change the filename to baseoutput (without camel case)?

mspaeth replied:
It is already changed, i dont know why git doesn't show it!

mbohlaender commented on v1/lib/baseOutput.php:
```diff
@@ -0,0 +1,23 @@
+ */
+
+/**
+ * This abstract class is the baseClass for the output plugins.
```
base class for all output plugins.

mbohlaender commented on v1/lib/baseOutput.php:
```diff
@@ -0,0 +1,23 @@
+<?php
+/**
+ * @file
+ * @version 2.8
```
Are we in diloc? no, so please change to 0.1

mbohlaender commented on v1/lib/cell.php:
```diff
@@ -1,25 +1,28 @@
 <?php
+/**
+ * @file
+ * @version 2.8
```
0.1

mspaeth committed:
Add new location of vendor dir to .gitignore #GoL work 1m

mspaeth committed:
Change version name to 0.1 #GoL work 5m

mbohlaender commented on v1/lib/cell.php:
```diff
@@ -1,25 +1,28 @@
- *
- * Class of a single cell of the the gamefield!
+ * This class is a cell object, every instance of this class is one cell
+ * which is used in an array to create together the gamefield.
```
this both lines can be done in one line, because it is one sentence. But because I wrote you another sentence this can be removed anyway.

mbohlaender commented on v1/lib/cell.php:
```diff
@@ -39,10 +43,37 @@ public function getCoordY()
+    }
+
+    /**
+     * Sets cell dead.
```
Change state of cell to be dead.

mbohlaender commented on v1/lib/cell.php:
```diff
@@ -39,10 +43,37 @@ public function getCoordY()
     }
+
+    /**
+     * Sets cell alive.
```
Change state of cell to be alive.

mbohlaender commented on v1/lib/baseOutput.php:
```diff
@@ -0,0 +1,23 @@
+abstract class BaseOutput
+{
+    /**
+     * This function needs to be implemented in every plugin which extends from this class
```
"." at the end, this is necessary in our diloc code, for generating a documentation with doxygen. So it's better when you learn it right from the beginning.

mbohlaender commented on v1/lib/cell.php:
```diff
@@ -1,25 +1,28 @@
- * Time: 12:01
- *
- * Class of a single cell of the the gamefield!
+ * This class is a cell object, every instance of this class is one cell
```
This class represents a cell, which contains information the information about whether this cell is alive or dead.

mbohlaender commented on v1/lib/consoleOutput.php:
```diff
@@ -0,0 +1,49 @@
+
+/**
+ * This class implements the output function from BaseOutput and prints the gamefield every cycle
+ * to the console.
```
one line!

mbohlaender replied:
This class inherits from the BaseOutput class, so we need to implement the output() function here, which will print every game cycle the complete gamefield on a console.

mbohlaender commented on v1/lib/consoleOutput.php:
```diff
@@ -0,0 +1,49 @@
+     * This function is extended from the baseoutput class.
+     * It prints the gamefield to the console for the number of cycles given.
+     *
+     * @param GameFieldController $_gameFieldController Object of gamefieldcontroller.
```
Think about it again? What you need to write here? Right! You need to write what for this parameter is needed...!

mbohlaender commented on v1/lib/consoleOutput.php:
```diff
@@ -0,0 +1,49 @@
+        {
+            // Make columns
+            for($i=0; $i<$y; $i++)
+            {
```
for (...)
{ // make columns

will save a line and is more readable

mbohlaender commented on v1/lib/consoleOutput.php:
```diff
@@ -0,0 +1,49 @@
+            {
+                // Make rows
+                for($j=0; $j<$x; $j++)
+                {
```
for (...)
{ // make rows

will save a line and is more readable

mbohlaender commented on v1/lib/gamefield.php:
```diff
@@ -1,26 +1,51 @@
     }
 
     /**
+     * Creates the array of cells which is the gamefield.
```
remove "which is the gamefield", because we know already, that we are in gamefield context.

mbohlaender commented on v1/runGame.php:
```diff
@@ -0,0 +1,92 @@
+                $numCycles = $options->getOption('numCycles');
+
+                $gameFieldController = new GameFieldController($gameField);
+                $output = new ConsoleOutput();
```
and now we are getting damn generic:
Please write something like $outputClass = $options-getOption("c")."Output"; ($options-getOption("c") must return Console", than build build output like $output = new $outputClass();

Than you need to move all your output plugins (just ConsoleOutput so far) into lib/output/ and then you need to include all files of this directory:
http://stackoverflow.com/questions/14595072/include-all-directory-files-in-class

mspaeth committed:
Move consoleoutput.php to /lib/output and include plugins with glob in runGame.php #GoL work 20m

Also add missing dots to documentation and change some documentation,
rename consoleOutput.php to consoleoutput.

mspaeth committed:
Change documentation in consoleoutput.php and add a dot in gamefield.php and gamefield.php #GoL work 10m
